plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -0,0 +1,260 @@
+   // the default value and can be ignored.
+   // Regardless, `should_pause()` and `reason()` called on returned
+   // `paused_status` is correct according to specified `pause_check`.
+   enum class pause_check { producer, other, both };
+
+   // Can be called concurrently with all member functions except:
+   //    + set_vote_timeout
+   //
+   // Returns a struct that provides useful time information tracked by this class
+   // which, most importantly, determines whether production should be paused.
+   // To determine whether production should be paused or not, simply check the boolean value
+   // returned from the should_pause member function of the returned struct.
+   pause_status check_pause_status(fc::time_point current_time, pause_check check = pause_check::both) const {
+      pause_status status;
+
+      const auto threshold_time = fc::time_point(current_time).safe_add(negative_vote_timeout);
+
+      auto process_vote_timing =
+         [&](const vote_timing& vt, pause_status::received_times& rt) {
+         const auto s = vt.get_vote_timing_status();
+
+         rt.latest_vote = s.latest_vote;
+
+         if (!s.first_block_after_vote.has_value())
+            return;
+
+         if (*s.first_block_after_vote < threshold_time) {
+            rt.earliest_conflict_block = *s.first_block_after_vote;
+         }
+      };
+
+      if (check != pause_check::producer) {
+         assert(check == pause_check::both || check == pause_check::other);
+         process_vote_timing(latest_other_vote, status.other_vote);
+      }
+      if (check != pause_check::other) {
+         assert(check == pause_check::both || check == pause_check::producer);
+         process_vote_timing(latest_producer_vote, status.producer_vote);
+      }
+
+      return status;
+   }
+
+private:
+   class vote_timing {
+   public:
+      vote_timing() : latest_vote(), first_block_after_vote() {}
+
+      // Can be called concurrently with all member functions.
+      bool record_received_vote(fc::time_point vote_received_time) {
+         auto vote_orig = latest_vote.load(std::memory_order_relaxed);
+
+         if (vote_received_time <= vote_orig)
+            return false;
+
+         latest_vote.store(vote_received_time, std::memory_order_relaxed);
+
+         return true;
+      }
+
+      // Can be called concurrently with all member functions.
+      bool record_received_block(fc::time_point block_received_time, fc::microseconds block_acceptance_delta) {
+         auto vote_time = latest_vote.load(std::memory_order_relaxed);
+
+         const auto adjusted_block_received_time = fc::time_point(block_received_time).safe_add(block_acceptance_delta);
+
+         if (adjusted_block_received_time <= vote_time)
+            return false;
+
+         auto block_orig = first_block_after_vote.load(std::memory_order_relaxed);
+
+         if ((vote_time < block_orig) && (block_orig <= block_received_time))
```

Comment:
Why this condition? Can you add a comment?

Reply:
Added some comments